File: samples/Sample.JavaScript/hmac-client.js
import crypto from 'crypto';

/**
 * HMAC Authentication utilities for JavaScript clients
 * Based on AspNetCore.HmacAuthentication implementation
 */
export class HmacClient {
    /**
     * @param {string} client - The client identifier
     * @param {string} secret - The secret key for HMAC signing
     * @param {string} baseUrl - The base URL of the API
     */
    constructor(client, secret, baseUrl = 'https://localhost:7134') {
        this.client = client;
        this.secret = secret;
        this.baseUrl = baseUrl.replace(/\/$/, ''); // Remove trailing slash

        // Constants from the .NET implementation
        this.DEFAULT_SCHEME_NAME = 'HMAC';
        this.TIME_STAMP_HEADER_NAME = 'x-timestamp';
        this.CONTENT_HASH_HEADER_NAME = 'x-content-sha256';
        this.EMPTY_CONTENT_HASH = '47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU=';
        this.DEFAULT_SIGNED_HEADERS = ['host', this.TIME_STAMP_HEADER_NAME, this.CONTENT_HASH_HEADER_NAME];
    }

    /**
     * Creates a canonical string for signing based on HTTP method, path, and header values
     * @param {string} method - HTTP method (GET, POST, etc.)
     * @param {string} pathAndQuery - Request path with query string
     * @param {string[]} headerValues - Array of header values in order
     * @returns {string} Canonical string for signing
     */
    createStringToSign(method, pathAndQuery, headerValues) {
        const upperMethod = method.toUpperCase();
        const headerString = headerValues.join(';');
        return `${upperMethod}\n${pathAndQuery}\n${headerString}`;
    }

    /**
     * Generates HMAC-SHA256 signature for the string to sign
     * @param {string} stringToSign - The canonical string to sign
     * @returns {string} Base64-encoded signature
     */
    generateSignature(stringToSign) {
        const hmac = crypto.createHmac('sha256', this.secret);
        hmac.update(stringToSign, 'utf8');
        return hmac.digest('base64');
    }

    /**
     * Generates the Authorization header value
     * @param {string[]} signedHeaders - Array of signed header names
     * @param {string} signature - Base64-encoded signature
     * @returns {string} Authorization header value
     */
    generateAuthorizationHeader(signedHeaders, signature) {
        const signedHeadersString = signedHeaders.join(';');
        return `${this.DEFAULT_SCHEME_NAME} Client=${this.client}&SignedHeaders=${signedHeadersString}&Signature=${signature}`;
    }

    /**
     * Calculates SHA256 hash of content and returns base64-encoded result
     * @param {string|null} content - Content to hash (null for empty content)
     * @returns {string} Base64-encoded SHA256 hash
     */
    calculateContentHash(content) {
        if (!content) {
            return this.EMPTY_CONTENT_HASH;
        }

        const hash = crypto.createHash('sha256');
        hash.update(content, 'utf8');
        return hash.digest('base64');
    }

    /**
     * Creates authenticated request headers
     * @param {string} method - HTTP method
     * @param {string} path - Request path with query string
     * @param {string|null} content - Request body content
     * @returns {Object} Headers object with authentication
     */
    createAuthenticatedHeaders(method, path, content = null) {
        const url = new URL(this.baseUrl + path);
        const host = url.host;
        const pathAndQuery = url.pathname + url.search;

        // Generate timestamp (Unix timestamp)
        const timestamp = Math.floor(Date.now() / 1000).toString();

        // Calculate content hash
        const contentHash = this.calculateContentHash(content);

        // Create header values in the order of DEFAULT_SIGNED_HEADERS
        const headerValues = [host, timestamp, contentHash];

        // Create string to sign
        const stringToSign = this.createStringToSign(method, pathAndQuery, headerValues);

        // Generate signature
        const signature = this.generateSignature(stringToSign);

        // Generate authorization header
        const authorizationHeader = this.generateAuthorizationHeader(this.DEFAULT_SIGNED_HEADERS, signature);

        // Build final headers
        const headers = {
            'Host': host,
            [this.TIME_STAMP_HEADER_NAME]: timestamp,
            [this.CONTENT_HASH_HEADER_NAME]: contentHash,
            'Authorization': authorizationHeader
        };

        // Add content-type for POST requests if not provided
        if (content) {
            headers['Content-Type'] = 'application/json';
        }

        return headers;
    }

    /**
     * Makes an authenticated HTTP request
     * @param {string} method - HTTP method
     * @param {string} path - Request path
     * @param {Object|null} body - Request body (will be JSON stringified)
     * @returns {Promise<Response>} Fetch response
     */
    async request(method, path, body = null) {
        const content = body ? JSON.stringify(body) : null;
        const headers = this.createAuthenticatedHeaders(method, path, content);

        const url = this.baseUrl + path;

        const requestOptions = {
            method: method.toUpperCase(),
            headers
        };

        if (content) {
            requestOptions.body = content;
        }

        // Import fetch dynamically for Node.js compatibility
        const fetch = (await import('node-fetch')).default;
        return fetch(url, requestOptions);
    }

    /**
     * Makes a GET request
     * @param {string} path - Request path
     * @returns {Promise<Response>} Fetch response
     */
    async get(path) {
        return this.request('GET', path, null);
    }

    /**
     * Makes a POST request
     * @param {string} path - Request path
     * @param {Object} body - Request body
     * @returns {Promise<Response>} Fetch response
     */
    async post(path, body) {
        return this.request('POST', path, body);
    }

    /**
     * Makes a PUT request
     * @param {string} path - Request path
     * @param {Object} body - Request body
     * @returns {Promise<Response>} Fetch response
     */
    async put(path, body) {
        return this.request('PUT', path, body);
    }

    /**
     * Makes a DELETE request
     * @param {string} path - Request path
     * @returns {Promise<Response>} Fetch response
     */
    async delete(path) {
        return this.request('DELETE', path, null);
    }
}

File: samples/Sample.Python/hmac_client.py
"""
HMAC Authentication utilities for Python clients
Based on HashGate.AspNetCore implementation
"""

import hashlib
import hmac
import base64
import time
from urllib.parse import urlparse, urlencode
from typing import Optional, Dict, Any, List
import requests


class HmacClient:
    """
    HMAC Authentication client for Python

    This client implements the same authentication scheme as HashGate.AspNetCore
    for making authenticated HTTP requests to APIs protected with HMAC authentication.
    """

    # Constants from the .NET implementation
    DEFAULT_SCHEME_NAME = "HMAC"
    TIME_STAMP_HEADER_NAME = "x-timestamp"
    CONTENT_HASH_HEADER_NAME = "x-content-sha256"
    EMPTY_CONTENT_HASH = "47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU="
    DEFAULT_SIGNED_HEADERS = ["host", TIME_STAMP_HEADER_NAME, CONTENT_HASH_HEADER_NAME]

    def __init__(self, client: str, secret: str, base_url: str = "https://localhost:7134"):
        """
        Initialize HMAC client

        Args:
            client: The client identifier
            secret: The secret key for HMAC signing
            base_url: The base URL of the API
        """
        self.client = client
        self.secret = secret.encode('utf-8')
        self.base_url = base_url.rstrip('/')  # Remove trailing slash
        self.session = requests.Session()

        # Disable SSL verification for development (localhost)
        if "localhost" in base_url or "127.0.0.1" in base_url:
            self.session.verify = False
            import urllib3
            urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

    def create_string_to_sign(self, method: str, path_and_query: str, header_values: List[str]) -> str:
        """
        Creates a canonical string for signing based on HTTP method, path, and header values

        Args:
            method: HTTP method (GET, POST, etc.)
            path_and_query: Request path with query string
            header_values: Array of header values in order

        Returns:
            Canonical string for signing
        """
        upper_method = method.upper()
        header_string = ";".join(header_values)
        return f"{upper_method}\n{path_and_query}\n{header_string}"

    def generate_signature(self, string_to_sign: str) -> str:
        """
        Generates HMAC-SHA256 signature for the string to sign

        Args:
            string_to_sign: The canonical string to sign

        Returns:
            Base64-encoded signature
        """
        signature = hmac.new(
            self.secret,
            string_to_sign.encode('utf-8'),
            hashlib.sha256
        ).digest()
        return base64.b64encode(signature).decode('utf-8')

    def generate_authorization_header(self, signed_headers: List[str], signature: str) -> str:
        """
        Generates the Authorization header value

        Args:
            signed_headers: Array of signed header names
            signature: Base64-encoded signature

        Returns:
            Authorization header value
        """
        signed_headers_string = ";".join(signed_headers)
        return f"{self.DEFAULT_SCHEME_NAME} Client={self.client}&SignedHeaders={signed_headers_string}&Signature={signature}"

    def calculate_content_hash(self, content: Optional[str]) -> str:
        """
        Calculates SHA256 hash of content and returns base64-encoded result

        Args:
            content: Content to hash (None for empty content)

        Returns:
            Base64-encoded SHA256 hash
        """
        if not content:
            return self.EMPTY_CONTENT_HASH

        content_bytes = content.encode('utf-8')
        hash_digest = hashlib.sha256(content_bytes).digest()
        return base64.b64encode(hash_digest).decode('utf-8')

    def create_authenticated_headers(self, method: str, path: str, content: Optional[str] = None) -> Dict[str, str]:
        """
        Creates authenticated request headers

        Args:
            method: HTTP method
            path: Request path with query string
            content: Request body content

        Returns:
            Headers dictionary with authentication
        """
        url = urlparse(self.base_url + path)
        host = url.netloc
        path_and_query = url.path + ("?" + url.query if url.query else "")

        # Generate timestamp (Unix timestamp)
        timestamp = str(int(time.time()))

        # Calculate content hash
        content_hash = self.calculate_content_hash(content)

        # Create header values in the order of DEFAULT_SIGNED_HEADERS
        header_values = [host, timestamp, content_hash]

        # Create string to sign
        string_to_sign = self.create_string_to_sign(method, path_and_query, header_values)

        # Generate signature
        signature = self.generate_signature(string_to_sign)

        # Generate authorization header
        authorization_header = self.generate_authorization_header(self.DEFAULT_SIGNED_HEADERS, signature)

        # Build final headers
        headers = {
            "Host": host,
            self.TIME_STAMP_HEADER_NAME: timestamp,
            self.CONTENT_HASH_HEADER_NAME: content_hash,
            "Authorization": authorization_header
        }

        # Add content-type for requests with content
        if content:
            headers["Content-Type"] = "application/json"

        return headers

    def request(self, method: str, path: str, json_data: Optional[Dict[str, Any]] = None) -> requests.Response:
        """
        Makes an authenticated HTTP request

        Args:
            method: HTTP method
            path: Request path
            json_data: Request body as dictionary (will be JSON serialized)

        Returns:
            requests.Response object
        """
        import json
        content = json.dumps(json_data) if json_data else None
        headers = self.create_authenticated_headers(method, path, content)

        url = self.base_url + path

        response = self.session.request(
            method=method.upper(),
            url=url,
            headers=headers,
            data=content
        )

        return response

    def get(self, path: str) -> requests.Response:
        """
        Makes a GET request

        Args:
            path: Request path

        Returns:
            requests.Response object
        """
        return self.request("GET", path, None)

    def post(self, path: str, json_data: Dict[str, Any]) -> requests.Response:
        """
        Makes a POST request

        Args:
            path: Request path
            json_data: Request body as dictionary

        Returns:
            requests.Response object
        """
        return self.request("POST", path, json_data)

    def put(self, path: str, json_data: Dict[str, Any]) -> requests.Response:
        """
        Makes a PUT request

        Args:
            path: Request path
            json_data: Request body as dictionary

        Returns:
            requests.Response object
        """
        return self.request("PUT", path, json_data)

    def delete(self, path: str) -> requests.Response:
        """
        Makes a DELETE request

        Args:
            path: Request path

        Returns:
            requests.Response object
        """
        return self.request("DELETE", path, None)
